### PR TITLE
implement FI_PEEK|CLAIM|DISCARD

### LINKED ISSUE #305: Implement FI_CLAIM/PEEK/DISCARD for trecvmsg path

### PR DESCRIPTION
**Pull request by jswaro**
closes #305 

@hppritcha @ztiffany @sungeunchoi @jshimek

### CODE REVIEW

File: prov/gni/src/gnix_ep.c [hunk 1]
```diff
@@ -542,7 +542,28 @@ static ssize_t gnix_ep_trecvmsg(struct fid_ep *ep,
...
-	clean_flags = (flags & GNIX_RECVMSG_FLAGS) | GNIX_MSG_TAGGED;
+	clean_flags = (flags & (GNIX_TRECVMSG_FLAGS)) | GNIX_MSG_TAGGED;
+
+	/* From the fi_tagged man page regarding the use of FI_CLAIM:
+	 *
+	 * In order to use the FI_CLAIM flag, an  application  must  supply  a
+     * struct  fi_context  structure as the context for the receive opera-
+     * tion.  The same fi_context structure used for an FI_PEEK + FI_CLAIM
+     * operation must be used by the paired FI_CLAIM request.
+	 */
```
Comment (ztiffany) [inline prov/gni/src/gnix_ep.c:553]: Odd indentation

File: prov/gni/src/gnix_ep.c [hunk 2]
```diff
@@ -542,7 +542,28 @@ static ssize_t gnix_ep_trecvmsg(struct fid_ep *ep,
...
+	if ((flags & FI_CLAIM) && _msg.context == NULL)
+		return -FI_EINVAL;
+
+	/* From the fi_tagged man page regaring the use of FI_DISCARD:
```
Comment (ztiffany) [inline prov/gni/src/gnix_ep.c]: regaring ->regarding

File: prov/gni/src/gnix_msg.c [hunk 3]
```diff
@@ -76,13 +86,7 @@ static int __gnix_recv_completion(struct gnix_fid_ep *ep,
...
 	assert(cq != NULL);
 
-	rc = _gnix_cq_add_event(cq,
-				req->user_context,
-				FI_RECV | FI_MSG,
-				req->msg.recv_len,
-				(void *)req->msg.recv_addr,
-				req->msg.imm,
-				0);
+	rc = _gnix_cq_add_event(cq, context, flags, len, addr, data, tag);
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:89]: The request supplied all the data in these parameter before.  Why change?  Is the request still needed?

Comment (jswaro) [inline prov/gni/src/gnix_msg.c:89]: The request was modified before providing the information. In the case of a peek, you don't want to modify the req, so a wrapper function from which the values could be passed in directly was made.

File: prov/gni/src/gnix_msg.c [hunk 4]
```diff
@@ -759,6 +908,14 @@ ssize_t _gnix_recv(struct gnix_fid_ep *ep, uint64_t buf, size_t len,
 			_gnix_fr_free(ep, req);
 		}
 	} else {
+		/* if peek/claim/discard, we didn't find what we
+		 * were looking for, return FI_ENOMSG
+		 */
+		if (r_flags) {
+			ret = -FI_ENOMSG;
+			goto pdc_exit;
+		}
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:943]: Don't we need to generate an error event in this case?

fi_tagged man page:
" If no message is found matching the tags specified in the peek
       request, then a completion queue error entry with err field set to FI_ENOMSG will be available."

File: prov/gni/src/gnix_msg.c [hunk 5]
```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
...
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "_gnix_ep_get_vc failed: %dn",
+				  ret);
+			return ret;
+		}
+
+		GNIX_INFO(FI_LOG_EP_DATA,
+				"returning rndzv completion for req, %p", req);
+
+
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c]: 2 blank lines was probably not intentional

File: prov/gni/src/gnix_msg.c [hunk 6]
```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
...
+		ret = _gnix_ep_get_vc(ep, src_addr, &req->vc);
+		if (ret) {
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "_gnix_ep_get_vc failed: %dn",
+				  ret);
+			return ret;
+		}
+
+		GNIX_INFO(FI_LOG_EP_DATA,
+				"returning rndzv completion for req, %p", req);
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:763]: GNIX_INFO lines in this function use inconsistent indentation.  One uses 2 tabs the other aligns with the open paren.

File: prov/gni/src/gnix_msg.c [hunk 7]
```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
...
+{
+	int ret;
+	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
+
+	GNIX_INFO(FI_LOG_EP_DATA, "discarding req, %p", req);
+
+	if (rendezvous) {
+		/* return a send completion so the sender knows the request/data
+		 * was sent, but discard the data locally
+		 */
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:745]: We return the RNDZV_FIN message here.  The term "send completion" seems overloaded.

File: prov/gni/src/gnix_msg.c [hunk 8]
```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
...
+		size_t len,
+		void *context,
+		uint64_t flags,
+		uint64_t tag,
+		uint64_t src_addr)
+{
+	int ret;
+	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
+
+	GNIX_INFO(FI_LOG_EP_DATA, "discarding req, %p", req);
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c]: I really appreciate these debug messages!

File: prov/gni/src/gnix_msg.c [hunk 9]
```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
...
+				"returning rndzv completion for req, %p", req);
+
+
+		/* send completion data. */
+		req->send_fn = __gnix_rndzv_req_send_comp;
+		ret = _gnix_vc_queue_req(req);
+	} else {
+		/* data has already been delivered, so just discard it */
+		ret = __recv_completion(ep, req, context, flags, len,
+				addr, req->msg.imm, tag);
```
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:773]: Is there any documentation that makes it clear that a completion event is generated from a discarded message?

Comment (jswaro) [inline prov/gni/src/gnix_msg.c:773]: A discard is always accompanied by either claim or peek. A successful peek should return a CQE, but not necessarily for a claim. I think there should be some open discussion on whether a discard should return a CQE at all, and I would think that it shouldn't. I'll remove it for now.

File: prov/gni/src/gnix_msg.c (hunk 4 above)
Comment (hppritcha) [inline prov/gni/src/gnix_msg.c:943]: I'm curious to about this one.   Why isn't a CQE being generated?

**Comment by ztiffany** [comment]
My understanding so far below.

Valid combinations and associated behaviors of these flags are:

PEEK - Generate success CQE if unclaimed, matching request is found, otherwise generate FI_ENOMSG error CQE
PEEK|CLAIM - Generate success CQE if unclaimed, matching request is claimed, otherwise generate FI_ENOMSG error CQE
PEEK|DISCARD - Generate success CQE if unclaimed, matching request is discarded, otherwise generate FI_ENOMSG error CQE
CLAIM|DISCARD - Generate success CQE if claimed, matching request is found and discarded, otherwise generate FI_ENOMSG error CQE

Other combinations result in error.

File: prov/gni/src/gnix_msg.c (hunk 4 above)
Comment (jswaro) [inline prov/gni/src/gnix_msg.c:943]: That should be outdated.

**Comment by jswaro** [comment]
> CLAIM|DISCARD - Generate success CQE if claimed, matching request is found and discarded, otherwise generate FI_ENOMSG error CQE

I'm not in agreement with that. FI_SUCCESS is return in the event that this discard is successful. A CQE typically means that data has landed successfully into the buffer locations given in the CQE, so that would be counter-intuitive. My understanding of what a CQE stands for may be skewed, so correct me if I'm wrong. The only reason we generate a CQE on PEEK|DISCARD is because it specifically states that a CQE must be generated on a successful peek, but I'm still not entirely sure that is accurate either, as it seems to follow the same reasoning as my previous argument.

**Comment by jswaro** [comment]
Can this be merged? Comments have been addressed

**Pull request merged**